### ISSUE TITLE
[Question] Best practices for profiling Equinox models

### ISSUE THREAD

**Issue by AakashKumarNain** [patrick-kidger/equinox#793]
I am building GPT kinda model in Equinox, and right now the forward pass is extremely slow compared to my torch implementation. I think this is one of the cases where I would like to attach a profiler and visualize the cost of every operation in my graph (model). 

Are there any recommended practices for using the profiler with equinox models? If there is no such guide, I am ready to contribute but will need some guidance

PS: I think it is high time we enable the `Discussions` in this repo 😄

**Comment by patrick-kidger** [comment]
Haha, I actually quite like using issues as discussions. Partly because half of people tend to use issues for that purpose anyway... ;)

So I think profiling Equinox should be the exact same as profiling JAX. `jax.profiler` is probably a natural place to start:

https://jax.readthedocs.io/en/latest/jax.profiler.html

I like the idea of creating some more docs on best practices here.

**Comment by AakashKumarNain** [comment]
Cool. Thanks a lot. I will push a PR once I get to run this exercise successfully on my end. If I encounter any issues, I will put them in this thread only.

**Comment by haydn-jones** [comment]
I've also struggled to get some of my models that use attention layers within them to perform as well as PyTorch. The performance gap I've seen has been up to 50% training throughput, but perhaps I'm not handling precision well enough. Would like to hear about what you find.

**Comment by patrick-kidger** [comment]
Usual warning to avoid including compile time if benchmarks, by the way.

FWIW I usually find the opposite -- PyTorch is usually slower than JAX.

Note that performance is a JAX-level thing, not an Equinox thing, so you'll probably find success by taking a look at some of the other JAX repos and seeing what tricks other people have.

**Comment by AakashKumarNain** [comment]
>Usual warning to avoid including compile time if benchmarks, by the way.

Yes, I am aware of it, and I generally warmup the model with a single batch

>FWIW I usually find the opposite -- PyTorch is usually slower than JAX.

That has been my experience for years, and this is why I am surprised by these numbers

> what tricks other people have.

The only tricks in this case can be `scan over for loop` for transformer layers, fused qkv, and efficient attention. I am pretty sure all are in place in my code

**Comment by AakashKumarNain** [comment]
Okay, I did some digging into this. I benchmarked each component separately, and found that all Equinox components are much faster than the torch counterparts. Beyond this, there are only two components for slowness:

1. Data loader -> Switched to `tf.data.Dataset` for now to resolve this. Though I wanted to avoid it in the first place as it creates another dependency
2. The major concern is `optax`. The forward pass is pretty efficient, but the backward pass is slow as hell. I don't know why it so slow but it instantly makes my JAX code ~1.5-2x slower than my torch code

Also, I have found that `XLA` flags for GPU either don't provide the speedups, or provide minor speedups. I will try this with H100, but not very hopeful. Will keep updating this thread

**Comment by Xynonners** [comment]
> Okay, I did some digging into this. I benchmarked each component separately, and found that all Equinox components are much faster than the torch counterparts. Beyond this, there are only two components for slowness:
> 
>     1. Data loader -> Switched to `tf.data.Dataset` for now to resolve this. Though I wanted to avoid it in the first place as it creates another dependency
> 
>     2. The major concern is `optax`. The forward pass is pretty efficient, but the backward pass is slow as hell. I don't know why it so slow but it instantly makes my JAX code ~1.5-2x slower than my torch code
> 
> 
> Also, I have found that `XLA` flags for GPU either don't provide the speedups, or provide minor speedups. I will try this with H100, but not very hopeful. Will keep updating this thread

since optax is the main large optimizer library in jax, I'm assuming you haven't found a solution to this?

the torch optimizers and jax optimizers seem to work very differently due to inplace operations.
